### ISSUE TITLE
PR

### ISSUE THREAD

**Issue by rkendel1** [rkendel1/browser-agent-runtime#4]
PR: Generic Local-LLM Semantic Decision Adapter

Objective

Make browser-agent-runtime the first concrete implementation of FeltDB’s runtime-neutral semantic-decision contract.

The goal is to demonstrate:

A normal local LLM can be exposed through a Jev-style typed decision interface without changing the underlying model.

This PR generalizes the existing WebLLM direct-logit implementation rather than creating a new semantic-decision system.

FeltDB remains authoritative for the semantic contract.

browser-agent-runtime owns model execution.

⸻

Architecture

FeltDB
  │
  │ Semantic Decision Runtime contract
  ▼
browser-agent-runtime
  │
  ├── LocalLLMDecisionRuntime
  │
  └── WebLLMAdapter
          │
          ▼
       WebLLM
          │
          ▼
      local model

The application should see:

Binary
Choice
Score
DecisionBatch

It should not need to know whether the underlying model is Qwen, Llama, Gemma, etc.

⸻

Scope

In scope

* Implement the generalized FeltDB DecisionRuntime.
* Refactor the existing WebLLM direct-logit runtime behind that interface.
* Add a model-neutral local LLM adapter boundary.
* Add deterministic candidate-label encoding.
* Support Binary and Choice through direct logits.
* Preserve existing Score behavior where already supported.
* Preserve DecisionBatch.
* Preserve provenance.
* Preserve context isolation.
* Add conformance tests.
* Add runtime capability reporting.
* Ensure the existing frozen browser benchmark can run through the new interface.

Out of scope

* Ollama.
* Transformers.js.
* Remote APIs.
* New models.
* Model training.
* Neural decision heads.
* Prompt optimization.
* Calibration.
* Automatic model routing.
* Eve integration.
* New semantic primitives.
* Changes to FeltDB semantics.
* Claiming compatibility with actual JEV.
* Ranking models.

⸻

1. Create the local model adapter boundary

Introduce an internal interface approximately equivalent to:

interface LocalLLMAdapter {
  readonly capabilities: LocalLLMCapabilities;
  execute(
    request: LocalLLMExecutionRequest
  ): Promise<LocalLLMExecutionResult>;
}

The adapter is responsible for model-specific execution.

It knows about:

* WebLLM
* tokenization
* logits
* model handles
* browser/WebGPU execution
* model-specific constraints

It does not define Binary/Choice/Score semantics.

⸻

2. WebLLM becomes an adapter

Refactor the existing WebLLM implementation into:

WebLLMAdapter

The existing direct-logit path becomes the first implementation of:

LocalLLMAdapter

No semantic behavior should change.

The current direct-logit behavior must remain:

decision request
      ↓
decision prompt
      ↓
candidate labels
      ↓
model logits
      ↓
restricted distribution
      ↓
selected option

⸻

3. Candidate-label encoding

Do not rely on arbitrary option strings being single model tokens.

Introduce deterministic label encoding.

For example:

Option 0 → A
Option 1 → B
Option 2 → C
Option 3 → D

The model sees:

A
B
C

while the runtime maintains:

A → "refund"
B → "replace"
C → "deny"

The mapping exists only for execution.

The returned semantic result remains:

"refund"

not:

"A"

Requirements

* deterministic
* stable within a decision
* unique
* bounded to supported labels
* validated before execution
* never persisted as semantic option identity

Add tests for:

* 2 options
* 3 options
* 4 options
* maximum supported options
* duplicate options
* invalid option sets

⸻

4. Direct-logit Choice

For a Choice request:

options = [
  "refund",
  "replace",
  "deny"
]

construct the runtime representation:

A = refund
B = replace
C = deny

Obtain the model’s candidate-token logits.

Calculate the distribution only over the permitted candidate labels.

Conceptually:

logits = [4.2, 2.7, 1.1]
softmax(logits)
       ↓
A: 0.80
B: 0.16
C: 0.04

Return:

result = "refund"
distribution = {
  refund: 0.80,
  replace: 0.16,
  deny: 0.04
}
selectedProbability = 0.80

Do not expose raw model token IDs as the semantic API.

⸻

5. Binary

Binary is implemented as a two-option Choice internally where appropriate:

A = true
B = false

or the equivalent existing direct-logit representation.

The public semantic result remains:

true | false

Do not create a second Boolean-specific model execution path unless WebLLM requires it.

⸻

6. Score

Preserve the existing Score implementation.

If Score is represented through discrete candidate labels, use the same candidate-label mechanism.

For example:

0
1
2
3
4
5

The runtime returns the declared score value.

Do not introduce continuous regression in this PR.

⸻

7. Generated-output fallback

Do not make this mandatory in the first WebLLM implementation.

However, structure the adapter boundary so a future runtime can report:

executionMethod:
  "logit"
  "structured"
  "generated"
  "head"

The WebLLM implementation should currently report:

executionMethod = "logit"

This is important because the long-term goal is:

Any local LLM
      ↓
Decision Adapter
      ↓
same typed decision interface

but different execution methods must remain distinguishable.

⸻

8. Capability reporting

The WebLLM adapter should expose capabilities equivalent to:

{
  binary: true,
  choice: true,
  score: true,
  batch: true,
  distribution: true,
  executionMethods: ["logit"]
}

Capabilities must describe reality.

Do not claim support for features that aren’t implemented.

⸻

9. DecisionBatch

Keep the existing DecisionBatch.

Refactor it so that:

DecisionBatch
      ↓
DecisionRuntime
      ↓
LocalLLMAdapter

rather than:

DecisionBatch
      ↓
WebLLM-specific implementation

Batch execution remains semantically independent of how the adapter optimizes execution.

The adapter may execute:

N independent calls

or eventually:

shared prefill
+ N decision readouts

without changing the semantic contract.

Do not implement shared-prefill/KV-cache optimization in this PR.

That should be a separate experiment after the frozen benchmark has actual observations.

⸻

10. Context handling

The adapter must receive the already-authorized context supplied by the semantic runtime.

It must not:

* fetch arbitrary application state
* perform authorization
* query FeltDB independently
* mutate durable state
* add hidden context
* silently retrieve additional documents

The execution boundary should be:

authorized decision request
          ↓
       adapter
          ↓
        model

This is important for proving that the model is not secretly getting information outside the decision contract.

⸻

11. Context isolation tests

Reuse the existing isolation fixtures.

Required cases:

Shared context

A + B both see shared fact

Both may use it.

Private A context

A sees private-A
B does not

Reordered decisions

A, B

versus:

B, A

must not associate results with the wrong decision.

Absent A

Removing A’s private context must not leave it available to B.

Irrelevant context

Adding irrelevant context must not cause cross-decision leakage.

These tests are semantic boundary tests, not model-quality tests.

⸻

12. Provenance

Every runtime result must preserve the FeltDB-required provenance and add browser-specific information where available.

Record:

runtime
runtime_version
model
model_revision
execution_method
prompt_revision
decision_schema_revision

Browser-specific fields:

browser
WebGPU availability
GPU/device
quantization
WebLLM version

Do not require a model SHA when the environment genuinely cannot provide one.

Record:

revision = unavailable

rather than inventing one.

This preserves the benchmark’s existing rule:

No silent substitution and no invented provenance.

⸻

13. Benchmark integration

The existing frozen benchmark becomes a consumer of the generic runtime.

It must continue to execute:

models:
  Qwen3-0.6B
  Qwen3.5-0.8B
  Qwen3.5-4B
runtimes:
  sequential
  DecisionBatch
  generated control

The model matrix remains frozen.

Do not modify the benchmark configuration in this PR.

The benchmark should now invoke:

DecisionRuntime

rather than WebLLM-specific functions.

⸻

14. Add a local-runtime smoke test

Create a minimal browser test that demonstrates:

Qwen/local model
        ↓
WebLLMAdapter
        ↓
LocalLLMDecisionRuntime
        ↓
FeltDB DecisionResult

For example:

Question:
"Should this invoice be escalated?"
Options:
["yes", "no"]

The test verifies:

* valid closed result
* distribution exists
* selected option is valid
* provenance exists
* execution method is logit

It must not assert a particular model answer unless the fixture is explicitly deterministic.

⸻

15. Do not test “intelligence” in unit tests

Avoid tests like:

expect(result).toBe("yes")

for ordinary semantic questions.

Those belong in the frozen benchmark.

Unit/conformance tests should test:

* shape
* contract
* isolation
* closed-set behavior
* provenance
* mapping
* execution correctness

The benchmark tests model quality.

This separation is critical.

⸻

16. Compatibility naming

Internally, it is acceptable to describe this as:

Jev-style semantic decision adapter

in documentation explaining the design.

Do not claim:

This is Jev.
This reproduces Jev.
This is equivalent to Jev.

The implementation is a generic closed-set LLM decision runtime inspired by the same class of interface.

The actual question being investigated is:

How much Jev-like decision behavior can ordinary local LLMs expose through a constrained decision interface?

That remains an empirical benchmark question.

⸻

17. Acceptance criteria

* [ ]	Existing WebLLM direct-logit runtime implements the generic local-model adapter.
* [ ]	FeltDB DecisionRuntime contract is consumed rather than duplicated.
* [ ]	Binary works.
* [ ]	Choice works.
* [ ]	Score remains supported where already implemented.
* [ ]	DecisionBatch works.
* [ ]	Candidate-label encoding is deterministic.
* [ ]	Candidate labels are not exposed as semantic option identities.
* [ ]	Closed-set validation remains enforced.
* [ ]	Context remains externally supplied and authorized.
* [ ]	No hidden state retrieval occurs.
* [ ]	Context isolation tests pass.
* [ ]	Provenance records execution method.
* [ ]	WebLLM reports logit execution.
* [ ]	Capability reporting reflects actual implementation.
* [ ]	Existing browser decision tests pass.
* [ ]	Frozen benchmark can execute through the generalized interface.
* [ ]	No model substitutions occur.
* [ ]	No new model-quality assertions are added to unit tests.
* [ ]	No shared-prefill/KV optimization is introduced.
* [ ]	No Eve integration is introduced.
* [ ]	No new semantic primitives are introduced.

⸻

Result

After this PR:

                    FeltDB
                       │
             SemanticDecision
                       │
               DecisionRuntime
                       │
                       ▼
          browser-agent-runtime
                       │
               LocalLLMRuntime
                       │
                 WebLLMAdapter
                       │
                       ▼
                  Local LLM

And the important property is now demonstrated in code:

Application
    │
    │ "make a Choice"
    ▼
FeltDB semantic contract
    │
    ▼
runtime-neutral adapter
    │
    ▼
Qwen / WebLLM

The model is no longer the interface.

The decision contract is the interface.

That is the foundation for adding Ollama or another local runtime later without changing ID8, Eve, or FeltDB’s semantic model.
